sema4ai/src/sema4ai_code/agents/collect_agent_spec_diagnostics.py

```diff
@@ -0,0 +1,226 @@
+
+if typing.TYPE_CHECKING:
+    from sema4ai_code.vendored_deps.yaml_with_location import ScalarInfo
+
+
+class _Node:
+    def __init__(self):
+        pass
+
+
+class _EntryNode(_Node):
+    def __init__(self, key: "ScalarInfo", value):
+        self.key = key
+        self.value = value
+
+
+class _ErrorNode(_Node):
+    def __init__(self, diagnostic):
+        self.diagnostic = diagnostic
+
+
+def _get_node(
+    data, key: str, msg: str, require_key_scalar_info: bool = True
+) -> _EntryNode | _ErrorNode:
+    from sema4ai_code.vendored_deps.yaml_with_location import (
+        ScalarInfo,
+        create_range_from_location,
+    )
+
+    report_location: Optional[tuple] = None
+    diagnostic: DiagnosticsTypedDict
+
+    if isinstance(data, _EntryNode):
+        report_location = data.key.location
+        data = data.value
+
+    report_location = report_location or (0, 0)
+
+    if not isinstance(data, dict):
+        diagnostic = {
+            "range": create_range_from_location(0, 0),
+            "severity": DiagnosticSeverity.Error,
+            "source": "sema4ai",
+            "message": msg,
+        }
+        return _ErrorNode(diagnostic)
+
+    if not require_key_scalar_info:
+        use_key = ScalarInfo(key, None)
+        ret = data.get(use_key, Sentinel.SENTINEL)
+        if ret != Sentinel.SENTINEL:
+            return _EntryNode(use_key, ret)
+    else:
+        for k, v in data.items():
+            # We do a (slow) search because we want the key as a ScalarInfo
+            # (for the position). We could pre-index so that it wasn't so
+            # slow, but I don't think it matters until we start to have
+            # really big `agent-spec.yaml` contents (which is not expected
+            # at this point).
+            #
+            # If the key position won't be used, `require_key_scalar_info`
+            # should be passed as False.
+            if isinstance(k, ScalarInfo):
+                if k.scalar == key:
+                    return _EntryNode(k, v)
+
+    diagnostic = {
+        "range": create_range_from_location(*report_location),
+        "severity": DiagnosticSeverity.Error,
+        "source": "sema4ai",
+        "message": msg,
+    }
+    return _ErrorNode(diagnostic)
+
+
+class _Analyzer:
+    def __init__(self, workspace: IWorkspace, doc: IDocument, monitor: IMonitor):
+        self.workspace = workspace
+        self.doc = doc
+        self.monitor = monitor
+
+        self._load_errors: list[DiagnosticsTypedDict] = []
+        self._loaded_yaml = False
+        self._yaml_data: Optional[dict] = None
+
+    def _load_yaml(self) -> None:
+        from yaml.error import MarkedYAMLError
+
+        from sema4ai_code.vendored_deps.yaml_with_location import (
+            LoaderWithLines,
+            ScalarInfo,
+            create_range_from_location,
+        )
+
+        if self._loaded_yaml:
+            return
+        self._loaded_yaml = True
```

Comment:
Is it intentional, that it's set before actual loading happens? Wouldn't it prevent the load from being retried if needed?

Reply:
If loading fails that's ok, it shouldn't be retried in the same analysis (a new analysis will then do a new load).